File: backend/src/main/java/com/agromercantil/truckmanagement/service/FipeService.java
package com.agromercantil.truckmanagement.service;

import java.util.Arrays;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import com.agromercantil.truckmanagement.model.fipe.FipeBrand;
import com.agromercantil.truckmanagement.model.fipe.FipeModel;
import com.agromercantil.truckmanagement.model.fipe.FipeInfo;
import com.agromercantil.truckmanagement.model.fipe.FipeYear;

@Service
public class FipeService {

        private static final String FIPE_API_URL = "https://fipe.parallelum.com.br/api/v2";
        private static final String VEHICLE_TYPE = "trucks";

        @Autowired
        private WebClient.Builder webClientBuilder;

        public List<FipeBrand> getBrands() {
                String url = String.format("%s/%s/brands", FIPE_API_URL, VEHICLE_TYPE);
                return webClientBuilder.build()
                                .get()
                                .uri(url)
                                .retrieve()
                                .bodyToMono(FipeBrand[].class)
                                .map(Arrays::asList)
                                .block();
        }

        public List<FipeModel> getModels(String brandCode) {
                String url = String.format("%s/%s/brands/%s/models", FIPE_API_URL, VEHICLE_TYPE, brandCode);
                return webClientBuilder.build()
                                .get()
                                .uri(url)
                                .retrieve()
                                .bodyToMono(FipeModel[].class)
                                .map(Arrays::asList)
                                .block();
        }

        public List<FipeYear> getYears(String brandCode, String modelCode) {
                String url = String.format("%s/%s/brands/%s/models/%s/years", FIPE_API_URL, VEHICLE_TYPE, brandCode,
                                modelCode);
                return webClientBuilder.build()
                                .get()
                                .uri(url)
                                .retrieve()
                                .bodyToMono(FipeYear[].class)
                                .map(Arrays::asList)
                                .block();
        }

        public FipeInfo getFipeInfo(String brandCode, String modelCode, String yearCode) {
                String url = String.format(
                                "%s/%s/brands/%s/models/%s/years/%s",
                                FIPE_API_URL, VEHICLE_TYPE, brandCode, modelCode, yearCode);

                return webClientBuilder.build()
                                .get()
                                .uri(url)
                                .retrieve()
                                .bodyToMono(FipeInfo.class)
                                .block();
        }
}

File: frontend/src/app/services/fipe.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class FipeService {
  private readonly API_URL = 'http://localhost:8080/api/fipe';

  constructor(private http: HttpClient) { }
  getBrands(): Observable<any[]> {
    return this.http.get<any[]>(`${this.API_URL}/brands`);
  }

  getModels(brandCode: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.API_URL}/brands/${brandCode}/models`);
  }

  getYears(brandCode:string, modelCode: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.API_URL}/brands/${brandCode}/models/${modelCode}/years`);
  }

  getFipeInfo(brandCode: string, modelCode: string, yearCode: string): Observable<any> {
    const params = new HttpParams()
      .set('brand', brandCode)
      .set('model', modelCode)
      .set('year', yearCode);

    return this.http.get<any>(`${this.API_URL}/validate`, { params });
  }
}
